### PR TITLE
Trigger the container restart when configuration change

### PR DESCRIPTION
**Pull request by rebtoor**
- added `_create_checksum` function in `container_config_hash` ansible module, to generate hashes for given folders of configuration files. This function was performed in a bash script called `container-puppet.sh`
- removed debug and step parameters from `container_config_hash` ansible module
- changed `edpm_container_manage_update_config_hash` boolean variable to true in edpm_container_manage role in order to enable hashes updates.

### CODE REVIEW

**Comment by rebtoor** [comment]
I created this PR as a draft because i'm unsure about a pair of things:

1. I replicated the generation of the md5sum following the [script|https://opendev.org/openstack/tripleo-heat-templates/src/branch/master/common/container-puppet.sh] command. In that script a tar container with a folder and a list of exclusions was created, resetting its mtime files, then the tar file was 'extracted' to stdout and passed to a sed command (needed to strip some puppet comments AFAIK) and finally hashed by md5sum. The first two tar commands basically create a concatenated text file from a list of files that is evaluated via sed and then hashed. Do we still need this? Why we can't just concatenate the files and generate the md5sum of all of them without using tar? We don't even need to reset their mtime (md5sum of the same file created in different time is the same).
2. The list of exclusion indicated in the original shell script, is still valid for edpm-ansible? And, if some file shouldn't be parsed, the "list" should be maintained in the  module itself? Shouldn't be the container role (e.g. edpm_iscsid) that calls the module the owner of that list? 

Let me know what do you think about it, i think there's room for optimizations. @slagle @kajinamit @bshephar @rabi

**Comment by rebtoor** [comment]
BTW, i created a runner (`quay.io/ralfieri/openstack-ansibleee-runner:hash`) with this branch' code, and i ran the playbook two times: when the edpm-compute-0 node was deployed and after I changed the configuration in the playbook (the hostname of rabbitmq endpoints).

The variable `EDPM_CONFIG_HASH` for `nova_compute` container has changed after the first and the second executions. The hash for nova_compute is a list of hashes, separated by `-` character because it mounts volumes from 2 different impacted folders: `/var/lib/config-data/ansible-generated/iscsid` (unchanged) and `/var/lib/config-data/ansible-generated/nova_libvirt` (changed)

```
[root@edpm-compute-0 ~]# podman inspect -f json nova_compute | jq .[].Config.Env 
  "LANG=en_US.UTF-8",
  "EDPM_CONFIG_HASH=d41d8cd98f00b204e9800998ecf8427e-78cab918a5af3acf96d285935f6490d9",
  "KOLLA_CONFIG_STRATEGY=COPY_ALWAYS",
  "LIBGUESTFS_BACKEND=direct",
  "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
  "TERM=xterm",
  "container=oci",
  "HOME=/var/lib/nova",
  "HOSTNAME=edpm-compute-0"
]


[
  "LIBGUESTFS_BACKEND=direct",
  "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
  "TERM=xterm",
  "container=oci",
  "LANG=en_US.UTF-8",
  "EDPM_CONFIG_HASH=d41d8cd98f00b204e9800998ecf8427e-668689aeb2d7d9c4e708b75281993888",
  "KOLLA_CONFIG_STRATEGY=COPY_ALWAYS",
  "HOME=/var/lib/nova",
  "HOSTNAME=edpm-compute-0"
]
```

**Review by slagle — COMMENTED** [review]
This looks good. I would not worry about the excluded files or sed logic right now. The excluded file we might need to worry about it would be /etc/libvirt/passwd.db. Maybe we could add a way to list files to exclude from the config hashing in the container defs json. The module could then exclude files based on the container instead of hardcoded like they are in the old shell script.

For the sed, I think that would only apply if ansible adds comments to files. I know that the blockinfile module does that, but we are not using that module for any container configuration, so I think we're good there.

**Comment by rebtoor** [comment]
Hey @slagle , thanks for the review!

> This looks good. I would not worry about the excluded files or sed logic right now. The excluded file we might need to worry about it would be /etc/libvirt/passwd.db. Maybe we could add a way to list files to exclude from the config hashing in the container defs json. The module could then exclude files based on the container instead of hardcoded like they are in the old shell script.

Makes sense, I was thinking something like that. Maintaining the hash generating function like this, we only need to change the tarfile filter function (_tarfile_cleanup) to avoid files with certain names.

> For the sed, I think that would only apply if ansible adds comments to files. I know that the blockinfile module does that, but we are not using that module for any container configuration, so I think we're good there.

Ansible generated comments are "static" (IIRC the template module put `#Managed by Ansible` while puppet put something with timestamp) so I'm not worried since it will be always the same.

Any thought on dropping tar file generation entirely?

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:155]: can we use context manager so that we ensure close this file?

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:156]: ditto. we should use context manager

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:162]: I'd define this as global constant instead of variable within the method.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:262]: config_hashes can't be none and is always a list.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:146]: We have to pull the logic to add/exclude some files, I guess.

https://github.com/openstack/tripleo-heat-templates/blob/master/common/container-puppet.sh#L164
https://github.com/openstack/tripleo-heat-templates/blob/master/common/container-puppet.sh#L170

Probably we want these override-able so that each role can define own requirements.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:155]: I think we no longer need tar file as it was used to just join file contents, which can be implemented differently here.

**Review by kajinamit — CHANGES_REQUESTED** [review]
Thanks for starting this work. Please find my comments inline.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: i changed completely the approach, see new commits :)

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: removed completely config_hashes since it's useless with the new approach.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: 100% agree, see new commits.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: For what I saw, there's no roles, at least for those that we have imported here, who write their config outside `/var/lib/config-data/ansible-generated` Even `logrotate-crond` and `iscsid` have their folder there. What should we add then?

About exclusions, I implemented the logic inside the `_create_checksum` function, defining an `exclusion` list, now defaulted to `[]`, but it's not referenced in its call yet. 

This module is called by the edpm_container_standalone role without any parameter because it was designed to run on *all* container startup config definition found in `CONTAINER_STARTUP_CONFIG` (which is a constant that points to `/var/lib/edpm-config/container-startup-config`) Handling an exclusion list for each service/role means that:

1. we change the logic behind this entire module, calling it for each service with their own parameters.
2. we add a variable in the container definition json file that enumerate a list of files that need to be excluded from hashing.

Are we sure that we still need to exclude files? Even if those files contains passwords or they binaries (like the ceph keyring linked), should they have the same hash if aren't changed?

I'm waiting for your feedback. (cc @slagle for another opinion)

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: no needed anymore

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (slagle) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: I think option 2 is what I was referring to earlier. I think we could manage the exclusion list in the container json (startup_config_json), and then pass the list into _set_config_hash (and then _create_checksum). So, it would still work per container, without any change to how the module is called.

We likely still need file exclusion for /etc/libvirt/passwd.db at least.

Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: @slagle done. I'm searching in the `container_config_json` for an environment var called `EDPM_EXCLUDE_FROM_HASH`. The variable itself is a list of files separated by colon, just like the ${PATH} variable style. e.g. `EDPM_EXCLUDE_FROM_HASH=/etc/nova/nova.conf:/etc/libvirt/passwd.db`

**Comment by rebtoor** [comment]
@kajinamit @slagle can you take another look at this so we can merge it? :) Thanks!

**Review by slagle — APPROVED** [review]

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (bshephar) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:182]: The variable assignment is unnecessary here, we can just return `self._create_checksum(config_volume, exclusions)`

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (bshephar) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:173]: Variable isn't necessary here, can just return `total_hash.hexdigest()`.

**Review by bshephar — APPROVED** [review]
Few comments that can be addressed in a follow up PR. Happy to merge this one as is.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:239]: Current logic does not initialize old_config_hash and exclude_from_hash in case environment key does not exist, which causes reference without assignment.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:243]: I think we better put this out of environment (I mean, add the key to the same level as environment) because this is not really used by container itself.

**Review by kajinamit — CHANGES_REQUESTED** [review]

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:182]: ack

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:173]: ack

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:239]: uh, thanks!

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:243]: It was my first thought but I kept variables in the container environment as they was. Let me do some tests.

**Comment by raukadah** [comment]
https://logserver.rdoproject.org/30/30/6f35f21fa9fccf2eefb02b449650edae43e197ae/github-check/edpm-ansible-github-rdo-integration-centos-8-crc-singlenode-centos-9-external-compute/f321a44/logs/controller/pod/deploy-external-dataplane-compute-f65pm-logs.txt
```
TASK [edpm_container_manage : Update container configs with new config hashes] ***
[0;31mAn exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'edpm_config_hash'[0m
[0;31mfatal: [38.102.83.101]: FAILED! => {"changed": false, "module_stderr": "Shared connection to 38.102.83.101 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1676629918.98837-1355-260497113266536/AnsiballZ_container_config_hash.py\", line 107, in <module>\r\n    _ansiballz_main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1676629918.98837-1355-260497113266536/AnsiballZ_container_config_hash.py\", line 99, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/root/.ansible/tmp/ansible-tmp-1676629918.98837-1355-260497113266536/AnsiballZ_container_config_hash.py\", line 47, in invoke_module\r\n    runpy.run_module(mod_name='ansible.modules.container_config_hash', init_globals=dict(_module_fqn='ansible.modules.container_config_hash', _modlib_path=modlib_path),\r\n  File \"/usr/lib64/python3.9/runpy.py\", line 225, in run_module\r\n    return _run_module_code(code, init_globals, run_name, mod_spec)\r\n  File \"/usr/lib64/python3.9/runpy.py\", line 97, in _run_module_code\r\n    _run_code(code, mod_globals, init_globals,\r\n  File \"/usr/lib64/python3.9/runpy.py\", line 87, in _run_code\r\n    exec(code, run_globals)\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 264, in <module>\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 260, in main\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 90, in __init__\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 237, in _update_hashes\r\nKeyError: 'edpm_config_hash'\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}[0m

```

**Comment by rebtoor** [comment]
> https://logserver.rdoproject.org/30/30/6f35f21fa9fccf2eefb02b449650edae43e197ae/github-check/edpm-ansible-github-rdo-integration-centos-8-crc-singlenode-centos-9-external-compute/f321a44/logs/controller/pod/deploy-external-dataplane-compute-f65pm-logs.txt
> 
> ```
> TASK [edpm_container_manage : Update container configs with new config hashes] ***
> �[0;31mAn exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'edpm_config_hash'�[0m
> �[0;31mfatal: [38.102.83.101]: FAILED! => {"changed": false, "module_stderr": "Shared connection to 38.102.83.101 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1676629918.98837-1355-260497113266536/AnsiballZ_container_config_hash.py\", line 107, in <module>\r\n    _ansiballz_main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1676629918.98837-1355-260497113266536/AnsiballZ_container_config_hash.py\", line 99, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/root/.ansible/tmp/ansible-tmp-1676629918.98837-1355-260497113266536/AnsiballZ_container_config_hash.py\", line 47, in invoke_module\r\n    runpy.run_module(mod_name='ansible.modules.container_config_hash', init_globals=dict(_module_fqn='ansible.modules.container_config_hash', _modlib_path=modlib_path),\r\n  File \"/usr/lib64/python3.9/runpy.py\", line 225, in run_module\r\n    return _run_module_code(code, init_globals, run_name, mod_spec)\r\n  File \"/usr/lib64/python3.9/runpy.py\", line 97, in _run_module_code\r\n    _run_code(code, mod_globals, init_globals,\r\n  File \"/usr/lib64/python3.9/runpy.py\", line 87, in _run_code\r\n    exec(code, run_globals)\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 264, in <module>\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 260, in main\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 90, in __init__\r\n  File \"/tmp/ansible_container_config_hash_payload_nl2gi8aq/ansible_container_config_hash_payload.zip/ansible/modules/container_config_hash.py\", line 237, in _update_hashes\r\nKeyError: 'edpm_config_hash'\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}�[0m
> ```
> 
> Here is the container built by the job: quay.rdoproject.org/openstack-k8s-operators/openstack-ansibleee-runner:86bf25981bbfebfc9168601878af072e115781b9

yeah, i need to do some tests locally first. Let me put this PR on draft.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: ok, now `edpm_config_hash` and `edpm_exclude_from_hash` are top level fields just like `environment`.

**Comment by rebtoor** [comment]
@kajinamit @slagle @bshephar I should have solved your reviews' points, let me know if it might work now.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:26]: Please keep dictionary order. (eg. glob should be placed before json after fnmatch.) This is not allowed by lint which is not enabled yet.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:174]: this function looks redundant. Also, this does not 'set' config hash and get would be more appropriate naming

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:157]: IMO create indicates this creates a resource. Probably get or calculate would be better.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (kajinamit) [inline edpm_ansible/ansible_plugins/modules/container_config_hash.py:244]: the calculated hash should be part of environment so that the underlying edpm_container_manage and podman module cna detect the different hash.

https://github.com/openstack-k8s-operators/edpm-ansible/blob/main/edpm_ansible/ansible_plugins/modules/edpm_container_manage.py#L299-L300

You have to update tripleo_container_manage if you get the result out of the environment hash.

**Review by kajinamit — CHANGES_REQUESTED** [review]
This version looks better now, but left me leave some comments as I found a few things we still have to fix.

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:26]: ack

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:157]: ack

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py:174]: ack

File: edpm_ansible/ansible_plugins/modules/container_config_hash.py
Comment (rebtoor) [reply edpm_ansible/ansible_plugins/modules/container_config_hash.py]: we talked about it in chat, i misunderstood your original comment. I restore `EDPM_CONFIG_HASH` as `environment` variable.

**Review by kajinamit — APPROVED** [review]
LGTM. It might be useful to squash all commits into a single one before we merge this PR to keep git history clean.

**Comment by rebtoor** [comment]
> LGTM. It might be useful to squash all commits into a single one before we merge this PR to keep git history clean.

Indeed, let me wait for CI test to succeed and I'll merge it squashing the commits.

Thanks a lot!

**Comment by rebtoor** [comment]
recheck

**Pull request merged**
